[FLINK-13354] [docs] Add documentation for how to use blink planner (#9362)
apache/flink: docs/dev/table/common.md

* This will be replaced by the TOC
{:toc}

> Main Differences Between the Two Planners

Author: i think list is ok, because too many details need to explain

-----------------------------------------

1. The Blink planner does not support `BatchTableSource`, only `StreamTableSource`. Blink treats batch jobs a special case of streaming. As such, batch jobs will not be translated into `DateSet` programs but translated into `DataStream` programs, the same as the streaming jobs.

> 2. The Blink planner does not support `ExternalCatalog` which is deprecated.

Reviewer: ```suggestion
2. The Blink planner only support the brand new `Catalog` and does not support `ExternalCatalog` which is deprecated.
```

1. Optimization of the logical plan, 
2. Translation into a DataStream program.

The behavior of translating  a query is different for `TableEnvironment` and `StreamTableEnvironment`.

For `TableEnvironment`, A Table API or SQL query is translated when `TableEnvironment.execute()` is called, because `TableEnvironment` will optimize multiple-sinks into one DAG.

while for `StreamTableEnvironment`, A Table API or SQL query is translated when:

* a `Table` is emitted to a `TableSink`, i.e., when `Table.insertInto()` is called.
* a SQL update query is specified, i.e., when `TableEnvironment.sqlUpdate()` is called.
* a `Table` is converted into a `DataStream`.

Once translated, a Table API or SQL query is handled like a regular DataStream program and is executed when `TableEnvironment.execute()` or `StreamExecutionEnvironment.execute()` is called.

</div>
</div>

{% top %}

Integration with DataStream and DataSet API
-------------------------------------------

> Both Flink and Blink streaming jobs can integrate with the `DataStream` API. Only Flink batch job can integrate with the `DataSet API`, Blink batch job could not be combined with both.

Reviewer 2: Please do not call them Flink and Blink jobs, they are all Flink jobs but with different planner (old and new planner).

Reviewer 3: nit: legacy sounds better old

Reviewer 2: I agree. However, we have used "old" in the API (`.useOldPlanner()`) and some other [documentation pages](https://ci.apache.org/projects/flink/flink-docs-master/dev/table/).

Once translated, a Table API or SQL query is handled like a regular DataStream or DataSet program and is executed when `StreamExecutionEnvironment.execute()` or `ExecutionEnvironment.execute()` is called.

{% endhighlight %}
</div>

<div data-lang="Blink planner" markdown="1">
{% highlight java %}

> Table API and SQL queries are translated into [DataStream]({{ site.baseurl }}/dev/datastream_api.html) program whether their input is a streaming or batch input. A query is internally represented as a logical query plan and is translated in two phases: 

Reviewer 3: ```suggestion
Table API and SQL queries are translated into [DataStream]({{ site.baseurl }}/dev/datastream_api.html) programs whether their input is streaming or batch. A query is internally represented as a logical query plan and is translated in two phases: 
```

> 1. optimization of the logical plan, 

Reviewer 3: ```suggestion
1. Optimization of the logical plan
```

> 2. translation into a DataStream program.

Reviewer 3: ```suggestion
2. Translation into a DataStream program
```

A Table API or SQL query is translated when:

* a `Table` is emitted to a `TableSink`, i.e., when `Table.insertInto()` is called.
* a SQL update query is specified, i.e., when `TableEnvironment.sqlUpdate()` is called.
* a `Table` is converted into a `DataStream` (only for Blink stream job).

Once translated, a Table API or SQL query is handled like a regular DataStream program and is executed when `TableEnvironment.execute()` or `StreamExecutionEnvironment.execute()` is called.

{% endhighlight %}
</div>
</div>

{% top %}

Integration with DataStream and DataSet API
-------------------------------------------

> Both Flink stream job and Blink stream job could be integrated with `DataStream`, and only Flink batch job could be integrated with `DataSet`, Blink batch job could not be integrated with both.

Reviewer 3: ```suggestion
Both Flink and Blink streaming jobs can integrate with the `DataStream` API. Only Flink batch job can integrate with the `DataSet API`, Blink batch job could not be combined with both.
```

> **Note:** The `DataSet` discussed next is only for the Flink batch planner.

Reviewer 3: ```suggestion
**Note:** The `DataSet` API discussed below is only relevant for the Flink batch planner.
```

Query Optimization
------------------

<div class="codetabs" markdown="1">
<div data-lang="Flink planner" markdown="1">

Apache Flink leverages Apache Calcite to optimize and translate queries. The optimization currently performed include projection and filter push-down, subquery decorrelation, and other kinds of query rewriting. Flink planner does not yet optimize the order of joins, but executes them in the same order as defined in the query (order of Tables in the `FROM` clause and/or order of join predicates in the `WHERE` clause).

It is possible to tweak the set of optimization rules which are applied in different phases by providing a `CalciteConfig` object. This can be created via a builder by calling `CalciteConfig.createBuilder())` and is provided to the TableEnvironment by calling `tableEnv.getConfig.setPlannerConfig(calciteConfig)`. 

</div>

<div data-lang="Blink planner" markdown="1">

> The foundation of Apache Flink query optimization is Apache Calcite. In addition to apply Calcite in optimization, Blink planner also does a lot to enhance it.

Reviewer 3: > Apache Flink leverages Apache Calcite to optimize and translate queries in conjunction with a large number of Flink specific optimizations and enhancements.

> The foundation of Apache Flink query optimization is Apache Calcite. In addition to apply Calcite in optimization, Blink planner also does a lot to enhance it.

Reviewer 3: @dawidwys @godfreyhe 
```
Apache Flink leverages and extends Apache Calcite to perform sophisticated query optimization. 

The Blink planner begins by performing a series of rule and cost-based optimizations, including but not limited to:
    * Subquery decorrelation based on Apache Calcite 
    * Project pruning
    * Partition pruning
    * Filter push-down 
    * Sub-plan deduplication to avoid duplicate computation. 
    * Special subquery rewriting:
        * Converts IN and EXISTS into left semi-joins
        * Converts NOT IN and NOT EXISTS into left anti-join 
    * Optional join reordering
        * Enabled via `table.optimizer.join-reorder-enabled`  

{% note %} IN/EXISTS/NOT IN/NOT EXISTS are currently only supported in conjunctive conditions.

The optimizer makes intelligent decisions, based not only on the plan but also rich statistics available from the data sources and fine-grain costs for each operator such as io, cpu, network, and memory. 

Advanced users may provide custom optimizations via a `CalciteConfig` that can be provided to the table environment by calling  `TableEnvironment#getConfig#setPlannerConfig`.
```

Fist of all, Blink planner does a series of rule-based optimization and cost-based optimization including:
* special subquery rewriting, including two part: 1. converts IN and EXISTS into left semi-join 2.converts NOT IN and NOT EXISTS into left anti-join. Note: only IN/EXISTS/NOT IN/NOT EXISTS in conjunctive condition is supported now.
* normal subquery decorrelation based on Apache Calcite
* projection pruning
* filter push down
* partition pruning
* join reorder if it is enabled (`table.optimizer.join-reorder-enabled` is true)
* other kinds of query rewriting

Secondly, Blink planner introduces rich statistics of data source and propagate those statistics up to the whole plan based on all kinds of extended `MetadataHandler`s. Optimizer could choose better plan based on those metadata.

Thirdly, Blink planner provides fine-grain cost of each operator, which takes io, cpu, network and memory into account. Cost-based optimization could choose better plan based on fine-grain cost definition.

Finally, Blink planner will try to find out duplicated sub-plans and reuse them to reduce duplicated computation. 

> It is possible to customize optimization programs referencing to `FlinkBatchProgram`(default optimization programs for batch) or `FlinkStreamProgram`(default optimization programs for stream), and replace the default optimization programs by providing a `CalciteConfig` object. This can be created via a builder by calling `CalciteConfig.createBuilder())` and is provided to the TableEnvironment by calling `tableEnv.getConfig.setPlannerConfig(calciteConfig)`. 

Reviewer 4: I find this paragraph quite unclear. What are the `FlinkBatchProgram` & `FlinkStreamProgram`? How can I use them? I would either add a complete new section on that (actually I wouldn't ;) ) or just refer to javadocs. Here I think it's enough to say that you can apply additional rules through `CalciteConfig`.

Query Optimization
------------------

<div class="codetabs" markdown="1">
<div data-lang="java" markdown="1">

> {% highlight Flink planner %}

Reviewer 3: Same as above, the highlights need to be removed since this is just plain text.

* join reorder if it is enabled (`table.optimizer.join-reorder-enabled` is true)
* other kinds of query rewriting

Secondly, Blink planner introduces rich statistics of data source and propagate those statistics up to the whole plan based on all kinds of extended `MetadataHandler`s. Optimizer could choose better plan based on those metadata.

Finally, Blink planner provides fine-grain cost of each operator, which takes io, cpu, network and memory into account. Cost-based optimization could choose better plan based on fine-grain cost definition .

It is possible to customize optimization programs referencing to `FlinkBatchProgram`(default optimization programs for batch) or `FlinkStreamProgram`(default optimization programs for stream), and replace the default optimization programs by providing a `CalciteConfig` object. This can be created via a builder by calling `CalciteConfig.createBuilder())` and is provided to the TableEnvironment by calling `tableEnv.getConfig.setPlannerConfig(calciteConfig)`. 

{% endhighlight %}
</div>
</div>

### Explaining a Table

The Table API provides a mechanism to explain the logical and optimized query plans to compute a `Table`. 

> This is done through the `TableEnvironment.explain(table)` method or `TableEnvironment.explain()` method. `explain(table)` returns the plan of given `Table`, `explain()` returns the result of multiple-sinks plan mainly used for Blink planner. It returns a String describing three plans: 

Reviewer 3: ```suggestion
This is done through the `TableEnvironment.explain(table)` method or `TableEnvironment.explain()` method. `explain(table)` returns the plan of a given `Table`. `explain()` returns the result of a multiple-sinks plan and is mainly used for the Blink planner. It returns a String describing three plans: 
```

table1.insertInto("MySink1");

Table table2 = table1.unionAll(tEnv.scan("MySource2"));
table2.insertInto("MySink2");

String explanation = tEnv.explain(false);
System.out.println(explanation);

{% endhighlight %}
</div>

<div data-lang="scala" markdown="1">
{% highlight scala %}
val settings = EnvironmentSettings.newInstance.useBlinkPlanner.inStreamingMode.build
val tEnv = TableEnvironment.create(settings)

val fieldNames = Array("count", "word")
val fieldTypes = Array[TypeInformation[_]](Types.INT, Types.STRING)
tEnv.registerTableSource("MySource1", new CsvTableSource("/source/path1", fieldNames, fieldTypes))
tEnv.registerTableSource("MySource2", new CsvTableSource("/source/path2",fieldNames, fieldTypes))
tEnv.registerTableSink("MySink1", new CsvTableSink("/sink/path1").configure(fieldNames, fieldTypes))
tEnv.registerTableSink("MySink2", new CsvTableSink("/sink/path2").configure(fieldNames, fieldTypes))

val table1 = tEnv.scan("MySource1").where("LIKE(word, 'F%')")
table1.insertInto("MySink1")

val table2 = table1.unionAll(tEnv.scan("MySource2"))
table2.insertInto("MySink2")

val explanation = tEnv.explain(false)
println(explanation)

{% endhighlight %}
</div>

> <div data-lang="python" markdown="1">

Reviewer 4: Why the python program differs so much? Shouldn't the python API be just a way to declare a relational query? I would assume the plan should be the same independent of which API we use.

Author: java code, scala code and python code is very similar for this example (explain multiple-sink plan), and their plans are also the same.  while the python plan is different from java & scala plan for above example (explain given table), because the python implementation of `from_elements` is different from java and scala: the python collection will be output a local file first, and then read as byte array in a `map`, finally deserialize the byte array to object list in a `flatMap` (the code is in `PythonBridgeUtils`) . so the `Execution Plan` about the source part is different.
